draco/analysis/transform.py

```diff
@@ -1087,31 +1106,206 @@ def process(self, polcont):
+
+            # If this is the weight dataset, keep track of where it is
+            # non-zero across all of the summed polarisations.
+            if name == weight_dset_name:
+                flag = mpiarray.ones(out_dset[:].global_shape, out_dset[:].axis)
 
-                sl = tuple([slice(None)] * pol_axis_pos)
-                out_dset[(*sl, 0)] = dset[(*sl, XX_ind)]
-                out_dset[(*sl, 0)] += dset[(*sl, YY_ind)]
+            # Loop over output polarisations
+            for oo, po in enumerate(self.pol):
+
+                oslc = make_slice(oo, pol_axis_pos)
+                pol_to_sum = self.P[po]
+                nsum = len(pol_to_sum)
+
+                # Loop over the input polarisations that we need to sum
+                for pi, sign in pol_to_sum.items():
+
+                    ii = input_pol.index(pi)
+                    islc = make_slice(ii, pol_axis_pos)
+
+                    if name == data_dset_name:
+                        # This is the primary data product, where we would like
+                        # to account for the sign.
+                        out_dset[oslc] += sign * dset[islc]
+
+                    elif name == weight_dset_name:
+                        # Keep track of what samples have non-zero weight
+                        # across all input polarisations.
+                        flag[oslc] &= dset[islc] > 0.0
+
+                        # Invert weights so that we properly propagate the variance.
+                        out_dset[oslc] += tools.invert_no_zero(dset[islc])
+
+                    elif np.issubdtype(out_dset.dtype, np.bool_):
+                        # All cases of boolean datasets with a polarisation axis
+                        # are masks, where a True value indicates a masked sample.
+                        # In this case, we will combine the masks, so that if any
+                        # of the input polarisations are masked then the output
+                        # polarisation is also masked.
+                        out_dset[oslc] |= dset[islc]
+
+                    else:
+                        # For all other datasets, we will simply take the average
+                        out_dset[oslc] += dset[islc]
+
+                # Normalize the output based on how many polarisations were summed
+                if name == weight_dset_name:
+                    out_dset[oslc] = (
+                        flag * nsum**2 * tools.invert_no_zero(out_dset[oslc])
+                    )
+
+                elif np.issubdtype(out_dset.dtype, np.integer):
+                    out_dset[oslc] //= nsum
+
+                elif np.issubdtype(out_dset.dtype, np.bool_):
+                    pass
 
-                if np.issubdtype(out_dset.dtype, np.integer):
-                    out_dset[:] //= 2
                 else:
-                    out_dset[:] *= 0.5
+                    out_dset[oslc] /= nsum
+
+        return outcont
+
+
+class PolWeightedAverage(task.SingleTask):
```

Comment:
If it isn't too difficult, I think we should combine this task with `transform.StokesIVis`, which is just doing an unweighted sum of XX and YY for vis and weights only.

Reply:
It’s not immediately clear to me what the best approach is here. One option might be to check whether a polarization axis exists and, if not, create one by selecting the appropriate subset of baselines. I’ve opened PR #348 to track this, but I plan to postpone further work until after the power spectrum analysis is finalized.